rust/kernel/init.rs

```diff
@@ -0,0 +1,827 @@
+//! API to safely and fallibly initialize pinned structs in-place.
+//!
+//! It also allows in-place initialization of big structs that would otherwise produce a stack overflow.
+
+use crate::{
+    error::{self, Error},
+    sync::UniqueArc,
+};
+use alloc::boxed::Box;
+use core::{
+    convert::Infallible,
+    marker::{PhantomData, Unpin},
+    mem::MaybeUninit,
+    pin::Pin,
+    ptr,
+};
+
+#[doc(hidden)]
+pub mod __private;
+
+/// Initialize a type on the stack. It will be pinned:
+/// ```rust
+/// # #![allow(clippy::blacklisted_name, clippy::new_ret_no_self)]
+/// # use kernel::{init, pin_init, stack_init, pin_data, init::*};
+/// # use core::pin::Pin;
+/// pin_data! {
+///     struct Foo {
+///         a: usize,
+///         b: Bar,
+///     }
+/// }
+///
+/// pin_data! {
+///     struct Bar {
+///         x: u32,
+///     }
+/// }
+///
+/// let a = 42;
+///
+/// stack_init!(let foo = pin_init!(Foo {
+///     a,
+///     b: Bar {
+///         x: 64,
+///     },
+/// }));
+/// let foo: Result<Pin<&mut Foo>> = foo;
+/// ```
+#[macro_export]
+macro_rules! stack_init {
+    (let $var:ident $(: $t:ty)? = $val:expr) => {
+        let mut $var = $crate::init::__private::StackInit$(::<$t>)?::uninit();
+        let val = $val;
+        let mut $var = unsafe { $crate::init::__private::StackInit::init(&mut $var, val) };
+    };
+}
+
+/// Construct an in-place initializer for structs.
+///
+/// The syntax is identical to a normal struct initializer:
+/// ```rust
+/// # #![allow(clippy::blacklisted_name, clippy::new_ret_no_self)]
+/// # use kernel::{init, pin_init, pin_data, init::*};
+/// # use core::pin::Pin;
+/// pin_data! {
```

Comment:
Should we use the convenience attribute in the examples? (i.e. users will typically use that one instead, right?)

Reply:
yes, forgot to add this, since I do not have the convenience attr in my crate